### DISCUSSION TITLE
Creating a new configuration with additional available tokens

### DISCUSSION THREAD

**Discussion by AnhTuanPW** [vladimir-bukhtoyarov/bucket4j#177] [Q&A] [answered]
Hi there,

I am using Bucket4J for my API rate limiting functionality. I have a use case where I need to replace a configuration but none of the existing `TokensInheritanceStrategy` strategies suit my needs. Is it possible to create new bandwidths with additional available tokens? 

Example:

Original configuration contains 100 tokens and 70 was used. As a result 30 tokens remains
New configuration set the capacity to 200 tokens so I want the bucket to have 130 tokens. 

In other words the formula is 
if (newCapacity - oldCapacity > 0)
newAvailableTokens = availableTokensBeforeReplacement + (newCapacity - oldCapacity) 

Most of methods that I need are private or protected in scope so I'm not sure if this is actually possible in a plausible way

**Comment by vladimir-bukhtoyarov** [comment]
Hello @AnhTuanPW

As you have already noticed, there is no functionality that can directly cover your use-case. But it looks that it should be a cheap change, I can release a new version at weekend. 

But I need some help from you to make this change: please propose a good name for constant there https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/TokensInheritanceStrategy.java as well as Javadocs.

**Reply by AnhTuanPW** [reply]
Thank you for your response. My proposal for the name of this strategy is `ADDITIVE`.

As for the JavaDoc:

> Instructs to copy available tokens as is, but with one exclusion: if new bandwidth capacity is greater than old capacity, available tokens will be increased by the difference between the old and the new configuration. 
> 
> The formula is newAvailableTokens = Math.min(availableTokensBeforeReplacement, newBandwidthCapacity) + Math.max(0, newBandwidthCapacity - capacityBeforeReplacement) 
> 
> Let's describe few examples.
> 
> Example 1: imagine bandwidth that was created by Bandwidth.classic(100, Refill.gready(10, Duration.ofMinutes(1))). At the moment of configuration replacement, it was 40 available tokens. After replacing this bandwidth by following Bandwidth.classic(200, Refill.gready(10, Duration.ofMinutes(1))) 40 available tokens will be copied and added to the difference between old and new configuration, and after replacement, we will have 140 available tokens.
> 
> Example 2: imagine bandwidth that was created by Bandwidth.classic(100, Refill.gready(10, Duration.ofMinutes(1))). At the moment of config replacement it was 40 available tokens. After replacing this bandwidth by following Bandwidth.classic(20, Refill.gready(10, Duration.ofMinutes(1)))), and after replacement we will have 20 available tokens.
>  
> Example 3: imagine bandwidth that was created by Bandwidth.classic(100, Refill.gready(10, Duration.ofMinutes(1))). At the moment of config replacement it was 10 available tokens. After replacing this bandwidth by following Bandwidth.classic(20, Refill.gready(10, Duration.ofMinutes(1)))), and after replacement we will have 10 available tokens.

**Reply by vladimir-bukhtoyarov** [reply]
Good. I am going to release new version on sunday.

**Comment by vladimir-bukhtoyarov** [comment]
> Most of methods that I need are private or protected in scope so I'm not sure if this is actually possible in a plausible way

It was explicit decision to not provide to the end-users the ability to write free form java code, because eventually it is planned to provide Redis Support based on Lua stored procedures, and it is obvious that it is not possible to translate java code into Lua.

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Has been released with version 6.3.0

**Reply by AnhTuanPW** [reply]
Thank you very much!
